editor/src/clj/editor/compute_program.clj

```diff
@@ -0,0 +1,177 @@
+
+(def ^:private form-data
+  (let [constant-values (protobuf/enum-values Material$MaterialDesc$ConstantType)]
+    {:navigation false
+     :sections
+     [{:title "Compute Program"
+       :fields
+       [{:path [:program]
+         :label "Compute Program"
+         :type :resource :filter "cp"}
+        {:path [:constants]
+         :label "Constants"
+         :type :table
+         :columns [{:path [:name] :label "Name" :type :string}
+                   {:path [:type]
+                    :label "Type"
+                    :type :choicebox
+                    :options (protobuf-forms/make-options constant-values)
+                    :default (ffirst constant-values)}
+                   {:path [:value] :label "Value" :type :vec4}]}
+        {:path [:samplers]
+         :label "Samplers"
+         :type :table
+         :columns (let [wrap-options (protobuf/enum-values Material$MaterialDesc$WrapMode)
+                        min-options (protobuf/enum-values Material$MaterialDesc$FilterModeMin)
+                        mag-options (protobuf/enum-values Material$MaterialDesc$FilterModeMag)]
+                    [{:path [:name] :label "Name" :type :string}
+                     {:path [:wrap-u]
+                      :label "Wrap U"
+                      :type :choicebox
+                      :options (protobuf-forms/make-options wrap-options)
+                      :default (ffirst wrap-options)}
+                     {:path [:wrap-v]
+                      :label "Wrap V"
+                      :type :choicebox
+                      :options (protobuf-forms/make-options wrap-options)
+                      :default (ffirst wrap-options)}
+                     {:path [:filter-min]
+                      :label "Filter Min"
+                      :type :choicebox
+                      :options (protobuf-forms/make-options min-options)
+                      :default (ffirst min-options)}
+                     {:path [:filter-mag]
+                      :label "Filter Mag"
+                      :type :choicebox
+                      :options (protobuf-forms/make-options mag-options)
+                      :default (ffirst mag-options)}
+                     {:path [:max-anisotropy]
+                      :label "Max Anisotropy"
+                      :type :number}])}]}]}))
+
+(defn- set-form-op [{:keys [node-id]} [property] value]
+  (g/set-property! node-id property value))
+
+(defn- clear-form-op [{:keys [node-id]} [property]]
+  (g/clear-property! node-id property))
+
+(g/defnk produce-form-data [_node-id name program constants samplers :as args]
+  (let [values (select-keys args (mapcat :path (get-in form-data [:sections 0 :fields])))
+        form-values (into {} (map (fn [[k v]] [[k] v]) values))]
+    (-> form-data
+        (assoc :values form-values)
+        (assoc :form-ops {:user-data {:node-id _node-id}
+                          :set set-form-op
+                          :clear clear-form-op}))))
```

Comment:
Couldn't this be wrapped in some sort of "property form data" helper to simplify the connection between forms and properties? I think I have seen this pattern of more or less generic set/clear functions + the produce function in multiple places, and perhaps it should be the default?

Reply:
True. I think the form view was originally used to edit `ini`-style settings files, and then repurposed for use with protobuf-based resources as well. Since then, the number of protobuf-based file formats that are edited using the form view have grown, and there is quite a bit of duplication. We could add a `editor/protobuf-form-util.clj` module where we put this stuff, perhaps?